notifier/notify_windows.go

```diff
@@ -1,16 +1,290 @@
+		return
+	}
+	initLogo()
+	snoreToastRunCmdAsync(exec.Command(notifierPath, "-id", n.GUID, "-t", "Portmaster", "-m", n.Message, "-b", buildSnoreToastButtonArgument(n.AvailableActions), "-appID", "io.safing.portmaster", "-p", logoLocation, "-pipeName", pipeName))
 }
 
 // Cancel cancels the notification.
 func (n *Notification) Cancel() {
+	if n == nil || n.GUID == "" {
+		return
+	}
+	snoreToastRunCmdAsync(exec.Command(notifierPath, "-close", n.GUID, "-appID", "io.safing.portmaster"))
+}
+
+//internal functions:
+
+func initNotifierPath() {
+	if notifierPath == "" {
+		var err error
+		notifierPath, err = getPath("notifier-snoretoast")
+
+		if err != nil {
+			log.Errorf("Error while getting SnoreToast-Path: %s %s", err, notifierPath)
+			return
+			//TODO: what to do?
+		}
+	}
+}
+
+func initLogo() {
+	if _, err := os.Stat(logoLocation); err != nil { //File doesn't exist or another Error that is handled while copying file there
+		os.MkdirAll(filepath.Dir(logoLocation), os.ModePerm)
+		if err = downloadFile(logoLocation, webLogoLocation); err != nil {
+			log.Errorf("Can't copy Logo for SnoreToast from %s to %s: %s", webLogoLocation, logoLocation, err)
+		}
+	}
+}
+
+func notificationListener() {
+	log.Debugf("Listening on Pipe %s\n", pipeName)
+	l, err := winio.ListenPipe(pipeName, nil)
+	if err != nil {
+		log.Errorf("Error while starting Pipe for SnoreToast: %s", err)
+		return
+	}
+	defer l.Close()
+
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			log.Errorf(err)
+			continue
+		}
+		go onCallback(conn)
+	}
+}
+
+func onCallback(conn net.Conn) {
+	data := snoreToastStrToMap(readWstrConnToStr(conn)) //TODO: what  if some required attribute isn't there?
+	log.Debugf("onCallback: %+v", data)
+
+	if data["action"] == "timedout" {
+		return
+	}
+
+	notificationsLock.Lock()
+	var n *Notification
+	for _, elem := range notifications {
+		if elem.GUID == data["notificationId"] {
+			n = elem
+		}
+	}
+	notificationsLock.Unlock()
+	if n == nil {
+		return
+	}
+	n.Lock()
+	defer n.Unlock()
+
+	switch data["action"] {
+	case "dismissed": //do nothing
+	case "clicked": //like dismissed: do nothing
+	case "buttonClicked":
+		for _, actions := range n.AvailableActions {
+			if actions.Text == data["button"] {
+				n.SelectAction(actions.ID)
+			}
+		}
+	default:
+		log.Debugf("SnoreToast-Action %s not implemented yet", data["action"])
+	}
+}
+
+//Snoretoast helper-Functions:
+
+func snoreToastStrToMap(in string) map[string]string {
+	ret := make(map[string]string)
+
+	for _, el := range strings.Split(in, ";") { //TODO: make sure, there is no ; in any value!
+		if el == "" {
+			continue
+		}
+
+		kv := strings.Split(el, "=") //TODO: make sure, there is no = in any value!
+		if len(kv) != 2 {
+			log.Warning("SnoreToast-Callback: " + el + " cannot be parsed in Key=Value")
+			continue
+		}
+
+		ret[kv[0]] = kv[1]
+	}
+
+	return ret
+}
+
+func buildSnoreToastButtonArgument(a []*Action) string {
```

Comment:
Hint: This function is perfect fit for attaching it to the `Notification` struct.